src/wallet/scriptpubkeyman.cpp

```diff
@@ -0,0 +1,597 @@
+    SetHDSeed(GenerateNewSeed(), force);
+    if (!NewKeyPool()) {
+        return false;
+    }
+    return true;
+}
+
+bool ScriptPubKeyMan::CanGenerateKeys()
+{
+    // A wallet can generate keys if it has an HD seed (IsHDEnabled) or it is a non-HD wallet (pre FEATURE_HD)
+    LOCK(wallet->cs_wallet);
+    return IsHDEnabled();
+}
+
+bool ScriptPubKeyMan::IsHDEnabled() const
+{
+    return !hdChain.IsNull();
+}
+
+bool ScriptPubKeyMan::CanGetAddresses(const uint8_t& type)
+{
+    LOCK(wallet->cs_wallet);
+    // Check if the keypool has keys
+    bool keypool_has_keys = false;
+    if (type == HDChain::ChangeType::INTERNAL && wallet->CanSupportFeature(FEATURE_HD_SPLIT)) {
+        keypool_has_keys = setInternalKeyPool.size() > 0;
+    } else if (type == HDChain::ChangeType::EXTERNAL) {
+        keypool_has_keys = KeypoolCountExternalKeys() > 0;
+    } else if (type == HDChain::ChangeType::STAKING) {
+        keypool_has_keys = setStakingKeyPool.size() > 0;
+    }
+    // If the keypool doesn't have keys, check if we can generate them
+    if (!keypool_has_keys) {
+        return CanGenerateKeys();
+    }
+    return keypool_has_keys;
+}
+
+size_t ScriptPubKeyMan::KeypoolCountExternalKeys()
+{
+    AssertLockHeld(wallet->cs_wallet);
+    return setExternalKeyPool.size() + set_pre_split_keypool.size();
+}
+
+unsigned int ScriptPubKeyMan::GetKeyPoolSize() const
+{
+    AssertLockHeld(wallet->cs_wallet);
+    return setInternalKeyPool.size() + setExternalKeyPool.size() + set_pre_split_keypool.size();
+}
+
+bool ScriptPubKeyMan::GetKeyFromPool(CPubKey& result, const uint8_t& changeType)
+{
+    if (!CanGetAddresses(changeType)) {
+        LogPrintf("%s: Cannot get address, type: %d\n", __func__ ,changeType);
+        return false;
+    }
+
+    CKeyPool keypool;
+    {
+        LOCK(wallet->cs_wallet);
+        int64_t nIndex;
+        if (!ReserveKeyFromKeyPool(nIndex, keypool, changeType)) {
+            if (wallet->IsLocked()) {
+                LogPrintf("%s: Wallet locked, cannot get address\n", __func__);
+                return false;
+            }
+            CWalletDB batch(wallet->strWalletFile);
+            result = GenerateNewKey(batch, changeType);
+            return true;
+        }
+        KeepDestination(nIndex);
+        result = keypool.vchPubKey;
+    }
+    return true;
+}
+
+bool ScriptPubKeyMan::GetReservedKey(const uint8_t& changeType, int64_t& index, CKeyPool& keypool)
+{
+    if (!CanGetAddresses(changeType)) {
+        return false;
+    }
+    if (!ReserveKeyFromKeyPool(index, keypool, changeType)) {
+        return false;
+    }
+    return true;
+}
+
+bool ScriptPubKeyMan::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool, const uint8_t& type)
+{
+    nIndex = -1;
+    keypool.vchPubKey = CPubKey();
+    {
+        LOCK(wallet->cs_wallet);
+
+        bool fReturningInternal = type == HDChain::ChangeType::INTERNAL;
+        fReturningInternal &= (IsHDEnabled() && wallet->CanSupportFeature(FEATURE_HD_SPLIT));
```

Comment:
isn't `wallet->CanSupportFeature(FEATURE_HD_SPLIT)` redundat here?

Reply:
yes, it is 👍 .